lib/src/query/filter.dart

```diff
@@ -1,35 +1,87 @@
 import 'dart:collection';
 
-class Filter with MapMixin<String, String> {
+class Filter with MapMixin<String, Object> {
   /// Example:
   /// ```dart
-  /// Filter({'post': '1,2', 'author': '12'}).addTo(url);
+  /// Filter({'post': '1,2', 'author': {'id': '12', 'role': 'admin'}}).addTo(url);
   /// ```
   /// encodes into
   /// ```
-  /// ?filter[post]=1,2&filter[author]=12
+  /// ?filter[post]=1,2&filter[author][id]=12&filter[author][role]=admin
   /// ```
-  Filter([Map<String, String> parameters = const {}]) {
+  Filter([Map<String, Object> parameters = const {}]) {
     addAll(parameters);
   }
 
-  static Filter fromUri(Uri uri) => Filter(uri.queryParametersAll
-      .map((k, v) => MapEntry(_regex.firstMatch(k)?.group(1) ?? '', v.last))
-    ..removeWhere((k, v) => k.isEmpty));
+  static Filter fromUri(Uri uri) {
+    Map<String, Object> filters = {};
+    uri.queryParametersAll.forEach((key, value) {
+      if (_validationRegex.hasMatch(key)) {
+        final matches = _extractionRegex.allMatches(key).toList();
+        _convertToMapAndMerge(matches, filters, value.last);
+      }
+    });
+    return Filter(filters);
+  }
+
+  static void _convertToMapAndMerge(
+    List<RegExpMatch> matches,
+    Map<String, Object> destination,
+    String value,
+  ) {
+    final key = matches[0].group(1) ?? '';
+    if (key.isNotEmpty) {
+      if (matches.length == 1) {
+        destination[key] = value;
+        return;
+      }
+      if (!destination.containsKey(key) ||
```

Comment:
Not sure how easy and clean it might be, but maybe try extracting `destination[key]` into a bar and utilize type check and inference to avoid force type conversion on line 44?